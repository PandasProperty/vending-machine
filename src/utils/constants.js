export const VENDING_MACHINE_CONFIG = {
    ROWS: 7,
    COLUMNS: 5,
    TOTAL_ITEMS: 15
};

export const VENDING_MACHINE_STATES = {
    IDLE: 'IDLE',
    READING: 'READING',
    PAYMENT: 'PAYMENT',
    DELIVER: 'DELIVER'
};

export const VENDING_MACHINE_CURRENCY = 'RON';

export const VENDING_MACHINE_MONEY = {
    0: {
        value: 0.5,
        label: '50 bani'
    },
    1: {
        value: 1,
        label: '1 leu'
    },
    2: {
        value: 5,
        label: '5 lei'
    },
    3: {
        value: 10,
        label: '10 lei'
    }
};