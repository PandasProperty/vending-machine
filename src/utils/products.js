
import { random, reduce, map, find } from 'lodash';
import {
    faHamburger,
    faPizzaSlice,
    faEgg,
    faCheese,
    faBreadSlice,
    faBacon,
    faHotdog,
    faIceCream,
    faFish,
    faCookie,
    faCandyCane,
    faCarrot,
    faLemon,
    faAppleAlt,
    faGlassWhiskey,
    faGlassMartini,
    faWineBottle,
    faBeer,
    faCoffee,
    faMugHot
} from '@fortawesome/free-solid-svg-icons';
import { VENDING_MACHINE_CONFIG } from './constants';

const colors = ['green', 'yellow', 'red', 'blue', 'purple', 'orange'];

const availableProducts = [{
    icon: faHamburger,
    name: 'Hamburger',
    price: 15
}, {
    icon: faPizzaSlice,
    name: 'Pizza',
    price: 15
}, {
    icon: faEgg,
    name: 'Kinder Egg',
    price: 5
}, {
    icon: faCheese,
    name: 'Cheese',
    price: 15
}, {
    icon: faBreadSlice,
    name: 'Toast',
    price: 2
}, {
    icon: faBacon,
    name: 'Bacon',
    price: 5
}, {
    icon: faHotdog,
    name: 'Hotdog',
    price: 7
}, {
    icon: faIceCream,
    name: 'Ice Cream',
    price: 5
}, {
    icon: faFish,
    name: 'Fish Chips',
    price: 15
}, {
    icon: faCookie,
    name: 'Cookie',
    price: 5
}, {
    icon: faCandyCane,
    name: 'Candy',
    price: 3
}, {
    icon: faCarrot,
    name: 'Carrot',
    price: 2
}, {
    icon: faLemon,
    name: 'Lemon',
    price: 3
}, {
    icon: faAppleAlt,
    name: 'Apple',
    price: 3
}, {
    icon: faGlassWhiskey,
    name: 'Whiskey',
    price: 20
}, {
    icon: faGlassMartini,
    name: 'Martini',
    price: 20
}, {
    icon: faWineBottle,
    name: 'Wine Bottle',
    price: 35
}, {
    icon: faBeer,
    name: 'Beer',
    price: 10
}, {
    icon: faCoffee,
    name: 'Coffee',
    price: 5
}, {
    icon: faMugHot,
    name: 'Hot Chocolate',
    price: 5
}];

export const generateVendingMachineContent = () => {
    const vendingMachineProducts = {};
    for (let rowIndex = 0; rowIndex < VENDING_MACHINE_CONFIG.ROWS; rowIndex++) {
        for (let columnIndex = 0; columnIndex < VENDING_MACHINE_CONFIG.COLUMNS; columnIndex++) {
            const productIndex = random(0, availableProducts.length - 1);
            const productId = rowIndex * VENDING_MACHINE_CONFIG.COLUMNS + columnIndex;
            vendingMachineProducts[productId] = {
                id: productId,
                ...availableProducts[productIndex],
                color: colors[random(0, colors.length - 1)],
                quantity: random(1, VENDING_MACHINE_CONFIG.TOTAL_ITEMS)
            };
        };
    };
    return vendingMachineProducts;
};

const computeCoins = (coins, length, ammount, max) => {
    if (!ammount) {
        return {
            coins,
            nrOfCoins: 0
        };
    }
    let rest = {
        coins,
        nrOfCoins: max
    };
    for (let i=0; i<length; i++) {
        if (coins[i].value <= ammount && coins[i].quantity > 0) {
            
            const newCoins = map(coins, coin => ({
                ...coin,
                quantity: coin.id === coins[i].id ? coin.quantity - 1 : coin.quantity
            }));

            let subRest = computeCoins(newCoins, length, ammount - coins[i].value, max);

            if (subRest.nrOfCoins !== max && subRest.nrOfCoins + 1 < rest.nrOfCoins) {
                rest = {
                    ...subRest,
                    nrOfCoins: subRest.nrOfCoins + 1
                };
                break;
            }
        }
    }
    return rest;
};

export const getMaximumChange = (coins, ammount) => {
    let max = 0;
    map(coins, coin => max += coin.quantity);
    return computeCoins(coins, coins.length, ammount, max);
};

export const getMachineCapital = (coins) => (
    reduce(coins, (sum, coin) => sum + coin.quantity * coin.value, 0)
);

export const getCoinsDescription = (oldCoins, newCoins) => {
    const description = [];
    map(oldCoins, oldCoin => {
        const newCoin = find(newCoins, newCoin => oldCoin.id === newCoin.id);
        description.push({
            ...newCoin,
            quantity: oldCoin.quantity - newCoin.quantity
        })
    });
    return description;
};
