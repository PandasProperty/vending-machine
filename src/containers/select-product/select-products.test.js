import React from 'react';
import { Provider } from 'react-redux';
import { faMugHot } from '@fortawesome/free-solid-svg-icons';
import { map } from 'lodash';
import SelectProducts from '.';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { VENDING_MACHINE_STATES, VENDING_MACHINE_CONFIG } from '../../utils/constants';

const mockStore = configureStore([]);

describe('Vending Machine', () => {
    const store = mockStore({
        vendingMachine: {
            products: map(
                Array(VENDING_MACHINE_CONFIG.ROWS * VENDING_MACHINE_CONFIG.COLUMNS), (item, index) => ({
                    color: 'orange',
                    icon: faMugHot,
                    id: index,
                    name: 'Cheese',
                    price: 15,
                    quantity: 7,
                })
            ),
            inputProduct: 1,
            mode: VENDING_MACHINE_STATES.PAYMENT,
            errorMessage: null,
        },
    });
    
    it('Select-products componet', () => {
        const component = renderer.create(
            <Provider store={store}>
                <SelectProducts />
            </Provider>
        );
        
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
