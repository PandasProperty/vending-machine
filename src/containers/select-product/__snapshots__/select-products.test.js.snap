// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vending Machine Select-products componet 1`] = `
<div
  className="select-product"
>
  <div
    className="display"
  >
    <div
      className="display__label"
    >
      Product
    </div>
    <div
      className="display__value"
    >
      1 - Cheese
    </div>
  </div>
  <div
    className="display"
  >
    <div
      className="display__label"
    >
      Price
    </div>
    <div
      className="display__value"
    >
      15 RON
    </div>
  </div>
  <div
    className="error__container"
  >
    
  </div>
  <div
    className="keypad"
  >
    <div
      className="keypad__row"
    >
      <div
        className="keypad__key"
        onClick={[Function]}
      >
        1
      </div>
      <div
        className="keypad__key"
        onClick={[Function]}
      >
        2
      </div>
      <div
        className="keypad__key"
        onClick={[Function]}
      >
        3
      </div>
    </div>
    <div
      className="keypad__row"
    >
      <div
        className="keypad__key"
        onClick={[Function]}
      >
        4
      </div>
      <div
        className="keypad__key"
        onClick={[Function]}
      >
        5
      </div>
      <div
        className="keypad__key"
        onClick={[Function]}
      >
        6
      </div>
    </div>
    <div
      className="keypad__row"
    >
      <div
        className="keypad__key"
        onClick={[Function]}
      >
        7
      </div>
      <div
        className="keypad__key"
        onClick={[Function]}
      >
        8
      </div>
      <div
        className="keypad__key"
        onClick={[Function]}
      >
        9
      </div>
    </div>
    <div
      className="keypad__row"
    >
      <div
        className="keypad__key keypad__key--enter"
        onClick={[Function]}
      >
        Enter
      </div>
      <div
        className="keypad__key"
        onClick={[Function]}
      >
        0
      </div>
      <div
        className="keypad__key keypad__key--cancel"
        onClick={[Function]}
      >
        Cancel
      </div>
    </div>
  </div>
</div>
`;
